raft/doc.go

```diff
@@ -180,5 +180,89 @@ cannot be removed any more since the cluster cannot make progress.
+	passes 'MsgHup' to its Step method and becomes candidate to start a new
+	election.
+
+	'MsgBeat' is for sending heartbeat. If a node is leader, tick function
+	in 'raft' struct is set as 'tickHeartBeat' that sends periodic heartbeat
+	messages in 'MsgBeat' type to its followers.
+
+	'MsgProp' proposes to append data to its log entries. This is a special
+	type to redirect proposals to leader. Therefore, send method overwrites
+	raftpb.Message's term with its HardState's term to not attach its local
+	term to 'MsgProp'. When 'MsgProp' is passed to leader's Step method,
+	leader first calls 'appendEntry' method to append entries to its log,
+	and then calls 'bcastAppend' method to send those entries to its peers.
+	When passed to candidate, 'MsgProp' is dropped. When passed to follower,
+	'MsgProp' is stored in follower's mailbox(msgs) by the send method. It is
+	stored with sender's ID and later forwarded to leader by rafthttp package.
+
+	'MsgApp' contains log entries to replicate. A leader calls bcastAppend,
+	which calls sendAppend, which sends soon-to-be-replicated logs in 'MsgApp'
+	type. When 'MsgApp' is passed to candidate's Step method, candidate reverts
+	back to follower, because it indicates that there is a valid leader sending
+	'MsgApp' messages. Candidate and follower respond to this message in
+	'MsgAppResp' type.
+
+	'MsgAppResp' is response to log replication request('MsgApp'). When
+	'MsgApp' is passed to candidate or follower's Step method, it responds by
+	calling 'handleAppendEntries' method, which sends 'MsgAppResp' to raft
+	mailbox.
+
+	'MsgVote' requests votes for election. When a node is follower or candidate
+	and 'MsgHup' is passed to its Step method, then the node call 'campaign'
+	method to campaign itself to become a leader. Once 'campaign' method is
+	called, the node becomes candidate and sends 'MsgVote' to followers
+	requesting votes. When passed to leader or candidate's Step method,
+	'MsgVote' is rejected ('MsgVoteResp' is returned with Reject true). When
+	'MsgVote' is passed to follower, it votes for the sender only when sender's
+	term is greater or sender's term is equal and sender's last committed index
+	is greater than or equal to follower's.
+
+	'MsgVoteResp' contains responses from voting request. When 'MsgVoteResp' is
+	passed to candidate, the candidate calculates how many votes it has won. If
+	it's more than majority (quorum), it becomes leader and calls 'bcastAppend'.
+	Otherwise, it reverts back to follower.
+
+	'MsgSnap' requests to install a snapshot message. When a node has just
+	become a leader or the leader receives 'MsgProp' message, it calls
+	'bcastAppend' method, which then calls 'sendAppend' method to each
+	follower. In 'sendAppend', if a leader fails to get term or entries,
+	the leader requests snapshot by sending 'MsgSnap' type message.
+
+	'MsgSnapStatus' tells the result of snapshot install message. When a
+	follower rejected 'MsgSnap', it indicates the snapshot request with
```

Comment:
This is not correct. MsgSnapStatus only reports failure if there was a network issue and the network layer failed to send out the snapshot to follower.